Add support for innergroup

utzig opened the pull request:
Doxygen groups that are defined inside another group, either by scope or by using \ingroup, are not automatically added to the documentation unless an explicit reference is created. This commit adds parsing of innergroup elements inside compound defs.

Fixes #555

utzig commented:
This is WIP, I am not so sure of how good a solution it is but it works for a project I am working on. A `no-inner` or some similar option might be interesting to have, similarly to `no-link`, to avoid the warnings from including groups multiple times. At least this happens in the documentation I am working on because we added explicit links to all innergroups...

utzig commented:
I'm not sure if I am missing something for a review to happen, but would be glad to get some feedback!

vermeeren commented:
I will try to review as soon as I can, a bit busy right now as a side effect of having been sick for a while. The same applies to other issues and PRs in the past week and a half or so. Sorry for the delay.

svenevs commented:
As a first pass it looks good to me, nice work / good addition! We need a test / example case for this, can you please add a simple innergroup example / make sure to include it in the sphinx docs? https://github.com/michaeljones/breathe/tree/master/examples

That will help us confirm it works as desired, and also show users how to use it. I would like to help review this further for @vermeeren (feel better soon! :heart:) but I don't really have any experience with doxygen groups... The example will help me confirm it works correctly :slightly_smiling_face:

svenevs commented on breathe/renderer/sphinxrenderer.py:
```diff
@@ -1024,6 +1024,12 @@ def render_derivedcompoundref(node):
         addnode('innernamespace', lambda: self.render_iterable(node.innernamespace))
 
+        if 'inner' in options:
+            for node in node.innergroup:
```
In the event that `inner` is requested but there are no inner groups, should we emit a warning along the lines of `inner groups requested for group {group name} but no inner groups found!`?

utzig replied:
Sure, and there's also the other way around. If there are `innergroup`s but no `inner` keyword was given in the options. To preserve compatibility this should obviously not trigger a warning. I am honestly not sure if someone would like the behavior of generating a warning in the case you pointed, but I will assume so and update accordingly. It can always be removed later! Btw, adding an `inner` in the documentation set I am working on, already triggers lots and lots of warnings because of the double references, because we basically had to create a reference to each group individually due to the lack of innergroup parsing.

utzig replied:
Thinking a bit better about this, turns out that `options` is part of a  "global" context in a sense. So if I have an innergroup B inside another group A, and I set `inner`, both A and B will be checked for the existence of innergroups. This will end up being an empty list for the innergroup B (assuming it has no innergroups itself!); but it should not trigger a warning.

utzig commented:
> As a first pass it looks good to me, nice work / good addition! We need a test / example case for this, can you please add a simple innergroup example / make sure to include it in the sphinx docs?

Thanks Stephen, I added a test: https://github.com/michaeljones/breathe/pull/556/commits/a127be6ce450fe6286df330ca7bcc0e5f5ff59af. An example already exists that creates an innergroup: https://github.com/michaeljones/breathe/blob/master/examples/doxygen/group.cpp#L26. It just so happens that it is currently ignored! I am not exactly sure what you mean by "sphinx docs", is adding this https://github.com/michaeljones/breathe/pull/556/commits/3e36e599f6f86c99e43c4bb79bcf619f4fac2933#diff-fb6911bd4f08815018ea2c65e04aa85cR45 not enough?

utzig commented:
> > As a first pass it looks good to me, nice work / good addition! We need a test / example case for this, can you please add a simple innergroup example / make sure to include it in the sphinx docs?
> 
> Thanks Stephen, I added a test: [a127be6](https://github.com/michaeljones/breathe/commit/a127be6ce450fe6286df330ca7bcc0e5f5ff59af). An example already exists that creates an innergroup: https://github.com/michaeljones/breathe/blob/master/examples/doxygen/group.cpp#L26. It just so happens that it is currently ignored! I am not exactly sure what you mean by "sphinx docs", is adding this [3e36e59#diff-fb6911bd4f08815018ea2c65e04aa85cR45](https://github.com/michaeljones/breathe/commit/3e36e599f6f86c99e43c4bb79bcf619f4fac2933#diff-fb6911bd4f08815018ea2c65e04aa85cR45) not enough?

I added an example with `:inner:` usage now, it was not clear to me what you meant before!

utzig commented:
@vermeeren Could you take a look?

vermeeren reviewed (approved):
Everything looks good to me @utzig , thanks for the patches!

utzig commented:
@vermeeren The official documentation does not yet reflect the changes, is it published automatically?

vermeeren commented:
@utzig There is a problem with the RTD docs being out of sync since quite a while, see #475. I will try asking Michael again and hopefully get a reply, I don't have permission on RTD I think.

If this isn't resolved soon I think I will self-host the breathe docs, because it is indeed quite a big problem.
